File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
class TrieNode{
public:
    vector<TrieNode*> children;
    bool isWord;
    TrieNode(){
        children.resize(26, NULL);
        isWord=false;
    }
};

class Trie {
private:
    TrieNode* root;
public:
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for(auto c:word){
            int idx = c-'a';
            if(node->children[idx]==NULL){
                node->children[idx] = new TrieNode();
            }
            node = node->children[idx];
        }
        node->isWord=true;
    }
    
    bool search(string word) {
        TrieNode* node = root;
        for(auto c:word){
            int idx = c-'a';
            if(node->children[idx]==NULL){
                return false;
            }
            node =node->children[idx];
        }
        return node->isWord==true;
    }
    
    bool startsWith(string prefix) {
        TrieNode* node = root;
        for(auto c:prefix){
            int idx = c-'a';
            if(node->children[idx]==NULL){
                return false;
            }
            node =node->children[idx];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.ts
class TrieNode{
    public eow: boolean;
    public children: TrieNode[];
    constructor(){
        this.eow=false;
        this.children=[];
        for(let i=0;i<26;i++){
            this.children.push(null);
        }
    }
}

class Trie {
    private root;
    constructor() {
        this.root = new TrieNode();
    }

    insert(word: string): void {
        let node = this.root;
        for(let i=0;i<word.length;i++){
            let idx: number = word[i].charCodeAt(0)-"a".charCodeAt(0);
            if(node.children[idx]==null){
                node.children[idx] = new TrieNode();
            }
            node = node.children[idx];
        }
        node.eow=true;
        
    }

    search(word: string): boolean {
        let node = this.root;
        for(let i=0;i<word.length;i++){
            let idx = word[i].charCodeAt(0) - "a".charCodeAt(0);
            if(node.children[idx]==null) return false;
            node = node.children[idx];
        }
        
        return node.eow;
    }

    startsWith(prefix: string): boolean {
        let node = this.root;
        for(let i=0;i<prefix.length;i++){
            let idx: number = prefix[i].charCodeAt(0) - "a".charCodeAt(0);
            if(node.children[idx]==null) return false;
            node = node.children[idx];
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */